compiler/rustc_passes/src/stability.rs

```diff
@@ -739,6 +739,28 @@ impl<'tcx> Visitor<'tcx> for Checker<'tcx> {
         intravisit::walk_poly_trait_ref(self, t);
     }
 
+    fn visit_use(&mut self, path: &'tcx UsePath<'tcx>, hir_id: HirId) {
+        let res = path.res;
+
+        // A use item can import something from two namespaces at the same time.
+        // For deprecation/stability we don't want to warn twice.
+        // This specifically happens with constructors for unit/tuple structs.
+        if let Some(ty_ns_res) = res.type_ns
+            && let Some(value_ns_res) = res.value_ns
+            && let Some(type_ns_did) = ty_ns_res.opt_def_id()
+            && let Some(value_ns_did) = value_ns_res.opt_def_id()
+            && let DefKind::Ctor(CtorOf::Struct, _) = self.tcx.def_kind(value_ns_did)
```

Comment:
```suggestion
            && let DefKind::Ctor(..) = self.tcx.def_kind(value_ns_did)
```
The same issue is relevant to enum variant constructors too.